WooCommerce/Classes/ViewRelated/Dashboard/Analytics Hub/AnalyticsHubTimeRangeController.swift

```diff
@@ -0,0 +1,146 @@
+import Foundation
+import Yosemite
+
+struct TimeRange {
+    let start: Date
+    let end: Date
+}
+
+public class AnalyticsHubTimeRangeController {
+
+    private let currentTimezone = TimeZone.current
+    private let currentDate: Date
+    private let currentCalendar: Calendar
+    private let selectionType: SelectionType
+
+    lazy private(set) var currentTimeRange: TimeRange = {
+        generateCurrentTimeRangeFrom(selectionType: selectionType)
+    }()
+
+    lazy private(set) var previousTimeRange: TimeRange = {
+        generatePreviousTimeRangeFrom(selectionType: selectionType)
+    }()
+
+    lazy private(set) var currentRangeDescription: String = {
+        generateDescriptionOf(timeRange: currentTimeRange)
+    }()
+
+    lazy private(set) var previousRangeDescription: String = {
+        generateDescriptionOf(timeRange: previousTimeRange)
+    }()
+
+    var selectionDescription: String {
+        selectionType.description
+    }
+
+    init(selectedTimeRange: StatsTimeRangeV4,
+         currentDate: Date = Date(),
+         currentCalendar: Calendar = Calendar(identifier: .iso8601)
+    ) {
+        self.selectionType = SelectionType.from(selectedTimeRange)
+        self.currentDate = currentDate
+        self.currentCalendar = currentCalendar
+    }
+
+    private func generateCurrentTimeRangeFrom(selectionType: SelectionType) -> TimeRange {
+        switch selectionType {
+        case .today:
+            return TimeRange(start: currentDate.startOfDay(timezone: currentTimezone), end: currentDate)
+        case .weekToDate:
+            let weekStart = currentDate.startOfWeek(timezone: currentTimezone, calendar: currentCalendar)
+            return TimeRange(start: weekStart, end: currentDate)
+        case .monthToDate:
+            return TimeRange(start: currentDate.startOfMonth(timezone: currentTimezone), end: currentDate)
+        case .yearToDate:
+            return TimeRange(start: currentDate.startOfYear(timezone: currentTimezone), end: currentDate)
+        }
+    }
+
+    private func generatePreviousTimeRangeFrom(selectionType: SelectionType) -> TimeRange {
+        switch selectionType {
+        case .today:
+            let oneDayAgo = currentCalendar.date(byAdding: .day, value: -1, to: currentDate)!
```

Comment:
I think we need to avoid using `implicitly unwrapped optionals` when dealing with dates.
It's probably better to return an optional, an error, or an estimated date that won't fail/crash.

Reply:
Done in https://github.com/woocommerce/woocommerce-ios/pull/8193/commits/bd37e9c8fb283f6b412b7354ac4e99ead6d5cefb